File: spare parts/fold26_production.c
/*
 * fold26_production.c
 * Production-ready C implementation of fold26 compression
 * Delta + RLE encoding for consensus log checkpoints
 *
 * Compile: gcc -O3 -Wall -o fold26 fold26_production.c -lcrypto
 * Usage: ./fold26 compress input.bin output.f26
 *        ./fold26 decompress input.f26 output.bin
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <openssl/sha.h>

#define FOLD26_VERSION "1.0.0"
#define FOLD26_MAGIC "F26\x01"
#define FOLD26_MAGIC_SIZE 4
#define FOLD26_HEADER_SIZE 40  /* magic(4) + size(4) + hash(32) */
#define RLE_ESCAPE 0xFF
#define MAX_RUN_LENGTH 255

typedef struct {
    uint8_t *data;
    size_t size;
    size_t capacity;
} Buffer;

typedef struct {
    size_t original_size;
    size_t compressed_size;
    float ratio;
    float reduction_pct;
    uint8_t hash[SHA256_DIGEST_LENGTH];
    int validated;
} CompressionMetadata;

/* Buffer management */
Buffer* buffer_create(size_t initial_capacity) {
    Buffer *buf = (Buffer*)malloc(sizeof(Buffer));
    if (!buf) return NULL;

    buf->data = (uint8_t*)malloc(initial_capacity);
    if (!buf->data) {
        free(buf);
        return NULL;
    }

    buf->size = 0;
    buf->capacity = initial_capacity;
    return buf;
}

void buffer_free(Buffer *buf) {
    if (buf) {
        free(buf->data);
        free(buf);
    }
}

int buffer_append(Buffer *buf, uint8_t byte) {
    if (buf->size >= buf->capacity) {
        size_t new_capacity = buf->capacity * 2;
        uint8_t *new_data = (uint8_t*)realloc(buf->data, new_capacity);
        if (!new_data) return -1;
        buf->data = new_data;
        buf->capacity = new_capacity;
    }

    buf->data[buf->size++] = byte;
    return 0;
}

int buffer_append_bytes(Buffer *buf, const uint8_t *bytes, size_t len) {
    for (size_t i = 0; i < len; i++) {
        if (buffer_append(buf, bytes[i]) != 0) return -1;
    }
    return 0;
}

/* Delta encoding */
Buffer* delta_encode(const uint8_t *data, size_t size) {
    if (size == 0) return buffer_create(0);

    Buffer *out = buffer_create(size);
    if (!out) return NULL;

    /* First byte stored as-is */
    buffer_append(out, data[0]);

    /* Encode deltas */
    for (size_t i = 1; i < size; i++) {
        uint8_t delta = (data[i] - data[i-1]) & 0xFF;
        buffer_append(out, delta);
    }

    return out;
}

/* Delta decoding */
Buffer* delta_decode(const uint8_t *data, size_t size) {
    if (size == 0) return buffer_create(0);

    Buffer *out = buffer_create(size);
    if (!out) return NULL;

    /* First byte stored as-is */
    buffer_append(out, data[0]);

    /* Reconstruct from deltas */
    for (size_t i = 1; i < size; i++) {
        uint8_t value = (out->data[i-1] + data[i]) & 0xFF;
        buffer_append(out, value);
    }

    return out;
}

/* RLE encoding */
Buffer* rle_encode(const uint8_t *data, size_t size) {
    if (size == 0) return buffer_create(0);

    Buffer *out = buffer_create(size);
    if (!out) return NULL;

    size_t i = 0;
    while (i < size) {
        uint8_t value = data[i];
        size_t count = 1;

        /* Count consecutive identical bytes */
        while (i + count < size && data[i + count] == value && count < MAX_RUN_LENGTH) {
            count++;
        }

        if (count >= 3) {
            /* Use RLE for runs >= 3 */
            buffer_append(out, RLE_ESCAPE);
            buffer_append(out, value);
            buffer_append(out, (uint8_t)count);
        } else if (count == 2) {
            /* Write pairs directly (cheaper than RLE overhead) */
            buffer_append(out, value);
            buffer_append(out, value);
        } else {
            /* Single byte */
            if (value == RLE_ESCAPE) {
                /* Escape 0xFF */
                buffer_append(out, RLE_ESCAPE);
                buffer_append(out, RLE_ESCAPE);
                buffer_append(out, 0x01);
            } else {
                buffer_append(out, value);
            }
        }

        i += count;
    }

    return out;
}

/* RLE decoding */
Buffer* rle_decode(const uint8_t *data, size_t size) {
    if (size == 0) return buffer_create(0);

    Buffer *out = buffer_create(size * 2);  /* Estimate */
    if (!out) return NULL;

    size_t i = 0;
    while (i < size) {
        if (data[i] == RLE_ESCAPE) {
            /* Escape sequence */
            if (i + 2 >= size) {
                fprintf(stderr, "Error: Truncated RLE escape at position %zu\n", i);
                buffer_free(out);
                return NULL;
            }

            uint8_t value = data[i + 1];
            uint8_t count = data[i + 2];

            /* Expand run */
            for (uint8_t j = 0; j < count; j++) {
                buffer_append(out, value);
            }

            i += 3;
        } else {
            /* Regular byte */
            buffer_append(out, data[i]);
            i++;
        }
    }

    return out;
}

/* Full compression */
Buffer* fold26_compress(const uint8_t *data, size_t size, CompressionMetadata *meta) {
    if (size == 0) {
        fprintf(stderr, "Error: Empty input\n");
        return NULL;
    }

    /* Calculate SHA-256 hash */
    SHA256(data, size, meta->hash);

    /* Stage 1: Delta encoding */
    Buffer *delta_buf = delta_encode(data, size);
    if (!delta_buf) {
        fprintf(stderr, "Error: Delta encoding failed\n");
        return NULL;
    }

    /* Stage 2: RLE */
    Buffer *rle_buf = rle_encode(delta_buf->data, delta_buf->size);
    buffer_free(delta_buf);
    if (!rle_buf) {
        fprintf(stderr, "Error: RLE encoding failed\n");
        return NULL;
    }

    /* Build output with header */
    size_t total_size = FOLD26_HEADER_SIZE + rle_buf->size;
    Buffer *output = buffer_create(total_size);
    if (!output) {
        buffer_free(rle_buf);
        return NULL;
    }

    /* Write header */
    buffer_append_bytes(output, (uint8_t*)FOLD26_MAGIC, FOLD26_MAGIC_SIZE);

    uint32_t size_le = (uint32_t)size;
    buffer_append_bytes(output, (uint8_t*)&size_le, sizeof(uint32_t));

    buffer_append_bytes(output, meta->hash, SHA256_DIGEST_LENGTH);

    /* Write compressed payload */
    buffer_append_bytes(output, rle_buf->data, rle_buf->size);

    buffer_free(rle_buf);

    /* Fill metadata */
    meta->original_size = size;
    meta->compressed_size = output->size;
    meta->ratio = (float)size / output->size;
    meta->reduction_pct = (1.0f - (float)output->size / size) * 100.0f;

    return output;
}

/* Full decompression */
Buffer* fold26_decompress(const uint8_t *data, size_t size, CompressionMetadata *meta) {
    if (size < FOLD26_HEADER_SIZE) {
        fprintf(stderr, "Error: Data too short for valid fold26 format\n");
        return NULL;
    }

    /* Validate magic */
    if (memcmp(data, FOLD26_MAGIC, FOLD26_MAGIC_SIZE) != 0) {
        fprintf(stderr, "Error: Invalid magic header\n");
        return NULL;
    }

    /* Parse header */
    uint32_t original_size;
    memcpy(&original_size, data + FOLD26_MAGIC_SIZE, sizeof(uint32_t));

    const uint8_t *stored_hash = data + FOLD26_MAGIC_SIZE + sizeof(uint32_t);

    const uint8_t *payload = data + FOLD26_HEADER_SIZE;
    size_t payload_size = size - FOLD26_HEADER_SIZE;

    /* Stage 2 (reverse): RLE decode */
    Buffer *rle_decoded = rle_decode(payload, payload_size);
    if (!rle_decoded) {
        fprintf(stderr, "Error: RLE decoding failed\n");
        return NULL;
    }

    /* Stage 1 (reverse): Delta decode */
    Buffer *output = delta_decode(rle_decoded->data, rle_decoded->size);
    buffer_free(rle_decoded);
    if (!output) {
        fprintf(stderr, "Error: Delta decoding failed\n");
        return NULL;
    }

    /* Validate hash */
    uint8_t computed_hash[SHA256_DIGEST_LENGTH];
    SHA256(output->data, output->size, computed_hash);

    if (memcmp(computed_hash, stored_hash, SHA256_DIGEST_LENGTH) != 0) {
        fprintf(stderr, "Error: Hash mismatch - data corrupted\n");
        buffer_free(output);
        return NULL;
    }

    /* Validate size */
    if (output->size != original_size) {
        fprintf(stderr, "Error: Size mismatch - expected %u, got %zu\n",
                original_size, output->size);
        buffer_free(output);
        return NULL;
    }

    /* Fill metadata */
    meta->original_size = original_size;
    meta->compressed_size = size;
    meta->validated = 1;
    memcpy(meta->hash, stored_hash, SHA256_DIGEST_LENGTH);

    return output;
}

/* File I/O */
int fold26_compress_file(const char *input_path, const char *output_path) {
    /* Read input file */
    FILE *fp = fopen(input_path, "rb");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open input file: %s\n", input_path);
        return -1;
    }

    fseek(fp, 0, SEEK_END);
    size_t size = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    uint8_t *data = (uint8_t*)malloc(size);
    if (!data) {
        fprintf(stderr, "Error: Memory allocation failed\n");
        fclose(fp);
        return -1;
    }

    fread(data, 1, size, fp);
    fclose(fp);

    /* Compress */
    CompressionMetadata meta = {0};
    Buffer *compressed = fold26_compress(data, size, &meta);
    free(data);

    if (!compressed) {
        return -1;
    }

    /* Write output */
    fp = fopen(output_path, "wb");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open output file: %s\n", output_path);
        buffer_free(compressed);
        return -1;
    }

    fwrite(compressed->data, 1, compressed->size, fp);
    fclose(fp);

    /* Print stats */
    printf("Compression successful:\n");
    printf("  Input: %s (%zu bytes)\n", input_path, meta.original_size);
    printf("  Output: %s (%zu bytes)\n", output_path, meta.compressed_size);
    printf("  Ratio: %.2fx\n", meta.ratio);
    printf("  Reduction: %.1f%%\n", meta.reduction_pct);

    buffer_free(compressed);
    return 0;
}

int fold26_decompress_file(const char *input_path, const char *output_path) {
    /* Read input file */
    FILE *fp = fopen(input_path, "rb");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open input file: %s\n", input_path);
        return -1;
    }

    fseek(fp, 0, SEEK_END);
    size_t size = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    uint8_t *data = (uint8_t*)malloc(size);
    if (!data) {
        fprintf(stderr, "Error: Memory allocation failed\n");
        fclose(fp);
        return -1;
    }

    fread(data, 1, size, fp);
    fclose(fp);

    /* Decompress */
    CompressionMetadata meta = {0};
    Buffer *decompressed = fold26_decompress(data, size, &meta);
    free(data);

    if (!decompressed) {
        return -1;
    }

    /* Write output */
    fp = fopen(output_path, "wb");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open output file: %s\n", output_path);
        buffer_free(decompressed);
        return -1;
    }

    fwrite(decompressed->data, 1, decompressed->size, fp);
    fclose(fp);

    /* Print stats */
    printf("Decompression successful:\n");
    printf("  Input: %s (%zu bytes)\n", input_path, size);
    printf("  Output: %s (%zu bytes)\n", output_path, meta.original_size);
    printf("  Hash: ");
    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        printf("%02x", meta.hash[i]);
    }
    printf("\n");

    buffer_free(decompressed);
    return 0;
}

/* Main */
int main(int argc, char *argv[]) {
    printf("fold26 Compressor v%s\n", FOLD26_VERSION);
    printf("Delta + RLE compression for consensus checkpoints\n\n");

    if (argc != 4) {
        printf("Usage:\n");
        printf("  Compress:   %s compress <input> <output>\n", argv[0]);
        printf("  Decompress: %s decompress <input> <output>\n", argv[0]);
        return 1;
    }

    const char *mode = argv[1];
    const char *input = argv[2];
    const char *output = argv[3];

    if (strcmp(mode, "compress") == 0) {
        return fold26_compress_file(input, output);
    } else if (strcmp(mode, "decompress") == 0) {
        return fold26_decompress_file(input, output);
    } else {
        fprintf(stderr, "Error: Unknown mode '%s'\n", mode);
        return 1;
    }
}

File: spare parts/fold26_production.py
#!/usr/bin/env python3
"""
Production-Ready fold26 Implementation
Delta + RLE compression with full encode/decode + validation
Ready for C porting
"""

import struct
import hashlib
import json
from typing import Tuple, Optional

class Fold26Compressor:
    """
    Production implementation of fold26 algorithm
    Combines delta encoding + run-length encoding
    Fully reversible with validation
    """

    VERSION = "1.0.0"
    MAGIC_HEADER = b'F26\x01'  # File magic for validation

    def __init__(self):
        self.stats = {
            'compressions': 0,
            'decompressions': 0,
            'bytes_in': 0,
            'bytes_out': 0,
            'errors': 0
        }

    def delta_encode(self, data: bytes) -> bytes:
        """
        Stage 1: Delta encoding
        Encodes differences between consecutive bytes
        """
        if len(data) == 0:
            return b''

        # First byte stored as-is
        out = bytearray([data[0]])

        # Encode deltas
        for i in range(1, len(data)):
            delta = (data[i] - data[i-1]) % 256
            out.append(delta)

        return bytes(out)

    def delta_decode(self, data: bytes) -> bytes:
        """
        Stage 1 (reverse): Delta decoding
        Reconstructs original from deltas
        """
        if len(data) == 0:
            return b''

        # First byte stored as-is
        out = bytearray([data[0]])

        # Reconstruct from deltas
        for i in range(1, len(data)):
            value = (out[-1] + data[i]) % 256
            out.append(value)

        return bytes(out)

    def rle_encode(self, data: bytes) -> bytes:
        """
        Stage 2: Run-length encoding
        Format: [value, count] for runs, [value] for singles
        Uses escape sequence for disambiguation
        """
        if len(data) == 0:
            return b''

        out = bytearray()
        i = 0

        while i < len(data):
            value = data[i]
            count = 1

            # Count consecutive identical bytes
            while i + count < len(data) and data[i + count] == value and count < 255:
                count += 1

            if count >= 3:
                # For runs >= 3, use RLE encoding
                # Format: [0xFF] [value] [count]
                out.append(0xFF)  # Escape marker
                out.append(value)
                out.append(count)
            elif count == 2:
                # For pairs, just write twice (cheaper than RLE overhead)
                out.append(value)
                out.append(value)
            else:
                # Single byte
                # If value is 0xFF, escape it
                if value == 0xFF:
                    out.append(0xFF)
                    out.append(0xFF)
                    out.append(0x01)  # Count of 1
                else:
                    out.append(value)

            i += count

        return bytes(out)

    def rle_decode(self, data: bytes) -> bytes:
        """
        Stage 2 (reverse): RLE decoding
        Reconstructs original from RLE format
        """
        if len(data) == 0:
            return b''

        out = bytearray()
        i = 0

        while i < len(data):
            if data[i] == 0xFF:
                # Escape sequence detected
                if i + 2 >= len(data):
                    raise ValueError(f"Truncated RLE escape at position {i}")

                value = data[i + 1]
                count = data[i + 2]

                # Expand run
                out.extend([value] * count)
                i += 3
            else:
                # Regular byte
                out.append(data[i])
                i += 1

        return bytes(out)

    def compress(self, data: bytes, validate: bool = True) -> Tuple[bytes, dict]:
        """
        Full fold26 compression pipeline
        Returns (compressed_data, metadata)
        """
        if len(data) == 0:
            return b'', {'error': 'Empty input'}

        original_size = len(data)
        original_hash = hashlib.sha256(data).digest()

        try:
            # Stage 1: Delta encoding
            delta_encoded = self.delta_encode(data)

            # Stage 2: RLE
            rle_encoded = self.rle_encode(delta_encoded)

            # Build header
            header = self.MAGIC_HEADER
            header += struct.pack('<I', original_size)  # Original size (4 bytes)
            header += original_hash  # SHA-256 hash (32 bytes)

            compressed = header + rle_encoded

            # Validation (optional but recommended)
            if validate:
                decompressed = self.decompress(compressed)[0]
                if decompressed != data:
                    raise ValueError("Validation failed: decompressed != original")

            # Update stats
            self.stats['compressions'] += 1
            self.stats['bytes_in'] += original_size
            self.stats['bytes_out'] += len(compressed)

            # Metadata
            metadata = {
                'original_size': original_size,
                'compressed_size': len(compressed),
                'ratio': original_size / len(compressed) if len(compressed) > 0 else 0,
                'reduction_pct': (1 - len(compressed)/original_size) * 100 if original_size > 0 else 0,
                'original_hash': original_hash.hex(),
                'validated': validate,
                'algorithm': 'fold26',
                'version': self.VERSION
            }

            return compressed, metadata

        except Exception as e:
            self.stats['errors'] += 1
            return b'', {'error': str(e)}

    def decompress(self, data: bytes, validate: bool = True) -> Tuple[bytes, dict]:
        """
        Full fold26 decompression pipeline
        Returns (decompressed_data, metadata)
        """
        if len(data) < 40:  # Minimum: magic(4) + size(4) + hash(32)
            return b'', {'error': 'Data too short for valid fold26 format'}

        try:
            # Parse header
            magic = data[:4]
            if magic != self.MAGIC_HEADER:
                return b'', {'error': f'Invalid magic header: {magic.hex()}'}

            original_size = struct.unpack('<I', data[4:8])[0]
            original_hash = data[8:40]

            # Extract compressed payload
            compressed_payload = data[40:]

            # Stage 2 (reverse): RLE decode
            rle_decoded = self.rle_decode(compressed_payload)

            # Stage 1 (reverse): Delta decode
            decompressed = self.delta_decode(rle_decoded)

            # Validation
            if validate:
                computed_hash = hashlib.sha256(decompressed).digest()
                if computed_hash != original_hash:
                    return b'', {'error': 'Hash mismatch: data corrupted'}

                if len(decompressed) != original_size:
                    return b'', {'error': f'Size mismatch: expected {original_size}, got {len(decompressed)}'}

            # Update stats
            self.stats['decompressions'] += 1

            metadata = {
                'original_size': original_size,
                'compressed_size': len(data),
                'decompressed_size': len(decompressed),
                'original_hash': original_hash.hex(),
                'validated': validate,
                'algorithm': 'fold26',
                'version': self.VERSION
            }

            return decompressed, metadata

        except Exception as e:
            self.stats['errors'] += 1
            return b'', {'error': str(e)}

    def compress_file(self, input_path: str, output_path: Optional[str] = None) -> dict:
        """Compress a file"""
        if output_path is None:
            output_path = input_path + '.f26'

        with open(input_path, 'rb') as f:
            data = f.read()

        compressed, metadata = self.compress(data)

        if 'error' in metadata:
            return metadata

        with open(output_path, 'wb') as f:
            f.write(compressed)

        metadata['input_path'] = input_path
        metadata['output_path'] = output_path

        return metadata

    def decompress_file(self, input_path: str, output_path: Optional[str] = None) -> dict:
        """Decompress a file"""
        if output_path is None:
            if input_path.endswith('.f26'):
                output_path = input_path[:-4]
            else:
                output_path = input_path + '.decompressed'

        with open(input_path, 'rb') as f:
            data = f.read()

        decompressed, metadata = self.decompress(data)

        if 'error' in metadata:
            return metadata

        with open(output_path, 'wb') as f:
            f.write(decompressed)

        metadata['input_path'] = input_path
        metadata['output_path'] = output_path

        return metadata

    def get_stats(self) -> dict:
        """Get compression statistics"""
        stats = self.stats.copy()

        if stats['bytes_in'] > 0:
            stats['overall_ratio'] = stats['bytes_in'] / stats['bytes_out'] if stats['bytes_out'] > 0 else 0
            stats['overall_reduction_pct'] = (1 - stats['bytes_out']/stats['bytes_in']) * 100
        else:
            stats['overall_ratio'] = 0
            stats['overall_reduction_pct'] = 0

        return stats

    def reset_stats(self):
        """Reset statistics"""
        self.stats = {
            'compressions': 0,
            'decompressions': 0,
            'bytes_in': 0,
            'bytes_out': 0,
            'errors': 0
        }


def test_fold26():
    """Comprehensive test suite for fold26"""
    print("="*80)
    print("FOLD26 PRODUCTION TEST SUITE")
    print("="*80)

    compressor = Fold26Compressor()

    # Test 1: Empty data
    print("\n📋 Test 1: Empty data")
    result = compressor.compress(b'')
    print(f"   Result: {'✓ PASS' if 'error' in result[1] else '✗ FAIL'}")

    # Test 2: Single byte
    print("\n📋 Test 2: Single byte")
    test_data = b'\x42'
    compressed, meta = compressor.compress(test_data, validate=True)
    if 'error' not in meta:
        decompressed, _ = compressor.decompress(compressed)
        print(f"   Original: 1 byte")
        print(f"   Compressed: {len(compressed)} bytes")
        print(f"   Ratio: {meta['ratio']:.2f}×")
        result = '✓ PASS' if decompressed == test_data else '✗ FAIL'
        print(f"   Result: {result}")
    else:
        print(f"   Result: ✗ FAIL - {meta['error']}")

    # Test 3: Repeated bytes (best case for RLE)
    print("\n📋 Test 3: Repeated bytes (100 × 0xAA)")
    data = b'\xAA' * 100
    compressed, meta = compressor.compress(data, validate=True)
    if 'error' not in meta:
        print(f"   Original: {len(data)} bytes")
        print(f"   Compressed: {len(compressed)} bytes")
        print(f"   Ratio: {meta['ratio']:.2f}×")
        print(f"   Reduction: {meta['reduction_pct']:.1f}%")
        print(f"   Result: ✓ PASS")
    else:
        print(f"   Result: ✗ FAIL - {meta['error']}")

    # Test 4: Sequential data (best case for delta)
    print("\n📋 Test 4: Sequential data (0-255)")
    data = bytes(range(256))
    compressed, meta = compressor.compress(data, validate=True)
    if 'error' not in meta:
        print(f"   Original: {len(data)} bytes")
        print(f"   Compressed: {len(compressed)} bytes")
        print(f"   Ratio: {meta['ratio']:.2f}×")
        print(f"   Reduction: {meta['reduction_pct']:.1f}%")
        print(f"   Result: ✓ PASS")
    else:
        print(f"   Result: ✗ FAIL - {meta['error']}")

    # Test 5: Random data (worst case)
    print("\n📋 Test 5: Random data (1000 bytes)")
    import random
    random.seed(42)
    data = bytes([random.randint(0, 255) for _ in range(1000)])
    compressed, meta = compressor.compress(data, validate=True)
    if 'error' not in meta:
        print(f"   Original: {len(data)} bytes")
        print(f"   Compressed: {len(compressed)} bytes")
        print(f"   Ratio: {meta['ratio']:.2f}×")
        print(f"   Reduction: {meta['reduction_pct']:.1f}%")
        print(f"   Result: ✓ PASS")
    else:
        print(f"   Result: ✗ FAIL - {meta['error']}")

    # Test 6: Time-series data (realistic)
    print("\n📋 Test 6: Time-series data (smooth changes)")
    value = 128
    data = bytearray()
    for i in range(1000):
        value = max(0, min(255, value + (i % 7) - 3))
        data.append(value)
    data = bytes(data)
    compressed, meta = compressor.compress(data, validate=True)
    if 'error' not in meta:
        print(f"   Original: {len(data)} bytes")
        print(f"   Compressed: {len(compressed)} bytes")
        print(f"   Ratio: {meta['ratio']:.2f}×")
        print(f"   Reduction: {meta['reduction_pct']:.1f}%")
        print(f"   Result: ✓ PASS")
    else:
        print(f"   Result: ✗ FAIL - {meta['error']}")

    # Test 7: Escape byte (0xFF) handling
    print("\n📋 Test 7: Escape byte handling (0xFF)")
    test_data = b'\xFF' * 50 + b'\x00' * 50
    compressed, meta = compressor.compress(test_data, validate=True)
    if 'error' not in meta:
        decompressed, _ = compressor.decompress(compressed)
        print(f"   Original: {len(test_data)} bytes")
        print(f"   Compressed: {len(compressed)} bytes")
        print(f"   Ratio: {meta['ratio']:.2f}×")
        result = '✓ PASS' if decompressed == test_data else '✗ FAIL'
        print(f"   Result: {result}")
    else:
        print(f"   Result: ✗ FAIL - {meta['error']}")

    # Test 8: Large data (10 MB)
    print("\n📋 Test 8: Large data (10 MB time-series)")
    import time
    value = 128
    data = bytearray()
    for i in range(10 * 1024 * 1024):
        value = max(0, min(255, value + ((i * 7) % 11) - 5))
        data.append(value)
    data = bytes(data)

    start = time.perf_counter()
    compressed, meta = compressor.compress(data, validate=False)  # Skip validation for speed
    compress_time = time.perf_counter() - start

    if 'error' not in meta:
        start = time.perf_counter()
        decompressed, _ = compressor.decompress(compressed, validate=False)
        decompress_time = time.perf_counter() - start

        print(f"   Original: {len(data):,} bytes")
        print(f"   Compressed: {len(compressed):,} bytes")
        print(f"   Ratio: {meta['ratio']:.2f}×")
        print(f"   Compress time: {compress_time:.3f}s ({len(data)/(1024*1024)/compress_time:.2f} MB/s)")
        print(f"   Decompress time: {decompress_time:.3f}s ({len(data)/(1024*1024)/decompress_time:.2f} MB/s)")
        result = '✓ PASS' if decompressed == data else '✗ FAIL'
        print(f"   Result: {result}")
    else:
        print(f"   Result: ✗ FAIL - {meta['error']}")

    # Statistics
    print("\n" + "="*80)
    print("OVERALL STATISTICS")
    print("="*80)
    stats = compressor.get_stats()
    print(f"Compressions: {stats['compressions']}")
    print(f"Decompressions: {stats['decompressions']}")
    print(f"Bytes in: {stats['bytes_in']:,}")
    print(f"Bytes out: {stats['bytes_out']:,}")
    print(f"Overall ratio: {stats['overall_ratio']:.2f}×")
    print(f"Overall reduction: {stats['overall_reduction_pct']:.1f}%")
    print(f"Errors: {stats['errors']}")

    print("\n✅ FOLD26 PRODUCTION TEST COMPLETE")


if __name__ == "__main__":
    test_fold26()
